org/mozilla/jss/pkcs11/PK11KeyGenerator.c

```diff
@@ -476,3 +477,62 @@ Java_org_mozilla_jss_pkcs11_PK11KeyGenerator_nativeClone
     return cloneObj;
 }
 
+/***********************************************************************
+ *
+ * PK11KeyGenerator.generateKBKDF
+ *
+ */
```

Comment:
Just wondering here, since from java land we have the ParameterSpec object available , why not just pass that instead of passing it AND a bunch of other members of the object separately???

Reply:
Some fields, such as `strength` are members of the `ParameterSpec`, but are separate parameters passed to `PK11_DeriveWithFlagsPerm`. Most of the other parameters (like the `KBKDFDataParameters`) get wrapped in an object and passed in the "mechanism parameters" (`param_item`) slot.

If we leave everything in the `ParameterSpec` object, this method looks like:

```c
generateKBKDF(parameterSpec) {
 int strength = getFieldFromJavaObject(parameterSpec, strength);
 long mechanism = getFieldFromJavaObject(parameterSpec, derivedKeyAlgorithm);
 ...
}
```

Each time you call `getFieldFromJavaObject`, you incur some overhead with the JNI call. Whereas if you pass them as arguments to `generateKBKDF` (like I am), Java handles all that setup for you, so theoretically the overhead should be better (and at worst, the same).

You also have a much longer method with more error handling logic. Java takes care of all that for you. So the sorter method, to me, wins. My 2c.